lib/urn.rb

```diff
@@ -1,26 +1,88 @@
+def URN(urn)
+  URN.new(urn)
+end
+
 class URN
+  InvalidURNError = Class.new(StandardError)
+
   PATTERN = %{(?i:urn:(?!urn:)[a-z0-9][a-z0-9\-]{1,31}:} +
-            %{(?:[a-z0-9()+,-.:=@;$_!*']|%[0-9a-f]{2})+)}.freeze
+            %{(?:[a-z0-9()+,-.:=@;$_!*']|%(?:2[1-9a-f]|[3-9a-f][0-9a-f]))+)}.freeze
   REGEX = /\A#{PATTERN}\z/
 
   attr_reader :urn
   private :urn
 
+  def self.extract(str)
+    if block_given?
+      str.scan(/#{PATTERN}/) { yield $& }
+      nil
+    else
+      result = []
+      str.scan(/#{PATTERN}/) { result.push $& }
+      result
+    end
```

Comment:
Both will have same result because the actual regular expression [has no groups](http://ruby-doc.org/core-2.3.0/String.html#method-i-scan), but in the future if we change the regular expression to have a group, this won't work.
I did that way following the example of Ruby core with their [URI.extract(str)](https://github.com/ruby/ruby/blob/trunk/lib/uri/rfc2396_parser.rb#L268) method.